### DISCUSSION TITLE
Keyboard macro and default M-x command selection

### DISCUSSION THREAD

**Discussion by yantar92** [minad/vertico#478] [Q&A] [answered]
Hi,

I am trying the following:

1. M-x butterfly RET (so that it comes up first in the completion list next time I invoke M-x)
2. F3 M-x RET yes RET F4
3. Switch to *sratch*
4. C-x C-k n test RET
5. M-x insert-kbd-macro
6. Observe 
    ```lisp
     (defalias 'test
   (kmacro "M-x <return> y e s <return>"))
   ```
7. M-x eval-buffer RET
8. M-x test
9. Observe execute-extended-command: ‘’ is not a valid command name

Interestingly, running the macro with F4 works fine.
I am wondering if it is something on vertico side or on Emac side.

**Comment by minad** [comment]
It is not clear to me what the problem is, but the repeated `M-x` invocations will mess up the history (`eval-buffer`, `test`). If I bind `'test` to a key, it works as well as F4 (assuming that `butterfly` is at the top).

**Reply by yantar92** [reply]
I can see history being messed up, but that should not cause empty command error. So, something else is going on there.

Ideally, it would be nice to have an option to record the actual command executed by M-x into a macro rather than just <RET> or navigation keys.

**Reply by minad** [reply]
I have problems to reproduce this due the aforementioned history changes. Such an error can occur if Vertico exits with an empty minibuffer in M-x. Then M-x complains about the invalid return value. This can happen in Vertico if the candidate list has not been loaded, and this has been the case in old Vertico versions.

However Vertico nowadays ensures that the candidate list is loaded via `vertico--prepare` which is installed as `pre-command-hook`:

https://github.com/minad/vertico/blob/1def56a1e678ff9a8eeafafda9c474a5fbb097c4/vertico.el#L626-L629
https://github.com/minad/vertico/blob/1def56a1e678ff9a8eeafafda9c474a5fbb097c4/vertico.el#L641

This means the candidates are loaded before the actual command execution, and the display is updated after the command (`vertico--exhibit`). Also the candidate list could be reloaded again afterwards in principle, e.g., if the input has changed. (Of course Vertico caches the candidate list, so no unnecessary recomputation should happen.)

So either you are running a very old Vertico version or something subtle is going on, which prevents `vertico--prepare` from doing its job.

**Reply by minad** [reply]
Maybe the `(and (symbolp this-command) (string-prefix-p "vertico-" (symbol-name this-command)))` condition in `vertico--prepare` is defeated by macro execution?

**Reply by yantar92** [reply]
> So either you are running a very old Vertico version or something subtle is going on, which prevents vertico--prepare from doing its job.

Well. What is really hapenning is that M-x test will change the command history, leaving "test" command on top. So,
M-x test will call "M-x return" that will select the most recent history entry, which is again "test", which ... (infinite loop continues for a while, until something is broken presumably in vertico).

Here is the recipe starting from emacs -Q:
1. emacs -Q
2. M-x package-install RET vertico RET
3. M-x vertico-mode
4. Eval  `(defalias 'test    (kmacro "M-x <return> y e s <return>"))`
5. M-x test

The above is not itself a problem though (from user perspective). The real problem is why M-x return ended up in the recorded macro to start with. When typing M-x RET, it would be more intuitive to somehow hook into kmacro to record the actual command being selected.

**Reply by minad** [reply]
There is no issue in your recipe. I get the following when evaluating it:

~~~
Error in post-command-hook (vertico--exhibit): (error "Lisp nesting exceeds ‘max-lisp-eval-depth’")
Error in pre-command-hook (vertico--prepare): (error "Lisp nesting exceeds ‘max-lisp-eval-depth’")
execute-extended-command: ‘’ is not a valid command name
~~~

You only managed to create an infinite kmacro loop which overflows the stack. The stack trace looks like this:

~~~ elisp
... many repeated macro invocations ...
  (funcall-interactively test 1)
  (command-execute test record)
  (execute-extended-command nil "test" nil)
  (funcall-interactively execute-extended-command nil "test" nil)
  (command-execute execute-extended-command)
  (execute-kbd-macro [134217848 return 121 101 115 return] 1 kmacro-loop-setup-function)
  (test 1)
  (funcall-interactively test 1)
  (command-execute test record)
  (execute-extended-command nil "test" nil)
  (funcall-interactively execute-extended-command nil "test" nil)
  (command-execute execute-extended-command)
  (execute-kbd-macro [134217848 return 121 101 115 return] 1 kmacro-loop-setup-function)
  (test 1)
... many repeated macro invocations ...
~~~

**Reply by yantar92** [reply]
> You only managed to create an infinite kmacro loop which overflows the stack

True.

> There is no issue in your recipe

I'd say that it is a usability issue from user perspective. Usually, macros work reliably and recording them does not require changes in the usual way we execute commands in Emacs. However, the normal vertico workflow with using navigation to select command does require special attention from users, which is surprising.

It is indeed not a bug, but something that would be nice to improve.

**Reply by minad** [reply]
No, I do not agree. To be clear - there is neither an issue, nor a "usability issue", nor a reliability issue, nor something to improve here. You can as easily create an infinitely looping kmacro with default completion as you can with Vertico:

~~~ elisp
(defalias 'test (kmacro "M-x t e s t <return>"))
~~~

**Reply by minad** [reply, marked as answer]
One thing we could maybe consider is to opt-out of Vertico at all when doing kmacro recording or execution. Maybe this is what you mean by improvement? This can be done in `vertico--setup` by checking `defining-kbd-macro` or `executing-kbd-macro`. Will this lead to a better behavior in your opinion?

**Reply by yantar92** [reply]
That sounds reasonable. I just checked source code of `execute-extended-command`, and is has huge chunks of the code skipped completely when `executing-kbd-macro` is non-nil. So, disabling a interactive features during macro execution is something Emacs itself does.

**Reply by minad** [reply]
Okay, let's do this then. During macro recording/execution we better want 100% deterministic behavior while during interactive use we want dynamic updates, sorting by recency/frecency and so on, so it makes sense to distinguish these scenarios.

**Reply by minad** [reply]
Done in https://github.com/minad/vertico/commit/b413071021a717bc25f730d2f94ce0740a9b8e34

**Reply by minad** [reply]
However now that I've implemented the change I am not so convinced anymore. If you type `M-x test RET` in Vertico it is just as deterministic as default completion. The only difference now is that the user interactivity is not available anymore, e.g., the RET in your recipe is disabled. Let's wait for user feedback and maybe revert.

**Reply by yantar92** [reply]
Hmm. Maybe disable history records while macro is recorded/executed?

**Reply by minad** [reply]
I am not sure. Disabling side effects may sound reasonable at first. However M-x during macro recording probably also modifies the minibuffer history?

**Reply by minad** [reply]
Well, by disabling Vertico as I have now, it is at least less suggestive to try to use interactive features. This makes the usage more reliable in macros. But for example if you press TAB in default completion in the minibuffer during macro expansion and then select a candidate via the arrow keys it will be just as problematic.

**Reply by yantar92** [reply]
> ... M-x during macro recording probably also modifies the minibuffer history?

It does, you are right. In fact, more complex macros may actually make use of minibuffer history. So, my idea is indeed not good.

**Reply by skat-kurt** [reply]
I was about to create a new issue, but just learned that this behavior was introduced on purpose. I would vote to revert these changes @minad . I 'd much prefer for vertico to be available when recording a keyboard macro. 

Unless I misunderstand something, as long as I would TAB to confirm my vertico selection during macro-recording things would be fine, no? I couldn't then rely on 'last selected' of course... 

The way it works now, where TAB brings up a new buffer that I have to switch to seems weird. If that new buffer contains many completion candidates I cannot consult-line search it, as I'm already in the minibuffer for the original completion.

**Reply by minad** [reply]
Let's wait for more feedback.

**Reply by Esnos33** [reply]
I also vote for reversion of changes or make this configurable. I don't think macros should be history based and you should always type out whole command and vertico help with this a a lot.

**Reply by minad** [reply]
@Esnos33 

> I don't think macros should be history based and you should always type out whole command and vertico help with this a a lot.

Yes, but the problem is that you have to remember to not use any other Vertico key bindings to get a reliable kmacro. You are only allowed to type out the command. Another idea could be to "lock" the Vertico minibuffer such that you cannot use the up/down keys. The result will be that the Vertico interaction will be restricted. The most extreme restriction is what we have now, we entirely disable Vertico.

However it could as well be that there are certain macros where you actually want to use the Vertico interactions. It may even make sense to create Vertico-specific macros, macros which only make sense when executed inside Vertico. Such uses are prevented if we disable Vertico. So I am probably going to revert the change.

**Reply by walseb** [reply]
I have disabled sorting in `M-x`, so that's a solution to get deterministic macros in the original scenario. I vote to have this change reverted or made optional. 

Below are some thoughts on the change:

I use a lot of macros, and although at times unpredictable, state changes caused by macros (like changes to candidate ordering, the buffer list, etc) is a feature that has helped me write more expressive macros numerous times. 

For instance, to repeat an action on a set of buffers, one can open up many buffers and record a macro for one buffer, then, bury/kill the buffer as the last action in the macro. Spamming the macro key then executes the macro in all those buffers. 

I have also frequently used the pattern of putting a string in the kill-ring, then pasting it to execute some relevant action in a Vertico menu (usually to overwrite a filename containing the string in the kill-ring, create a folder, or execute an action containing that string). 

While this change removes some of that unpredictability, it also makes those kinds of quick and dirty macros harder to write. Also, as I rely entirely on Vertico to navigate M-x, it becomes very difficult to run any M-x commands while recording a macro after this change.

**Reply by minad** [reply]
Yes, I agree. Reverted in https://github.com/minad/vertico/commit/ba650a7ab90d66686ba787937ac9e71f749c598e.
